cmd/clusterctl/client/cluster/objectgraph.go

```diff
@@ -614,17 +622,41 @@ func (o *objectGraph) setSoftOwnership() {
 		}
 	}
 
+	ownerClasses := map[*node]empty{}
 	clusterClasses := o.getClusterClasses()
+
 	// Cluster that uses a ClusterClass are soft owned by that ClusterClass.
 	for _, clusterClass := range clusterClasses {
 		for _, cluster := range clusters {
 			// if the cluster uses a managed topology and uses the clusterclass
 			// set the clusterclass as a soft owner of the cluster.
-			if className, ok := cluster.additionalInfo[clusterTopologyNameKey]; ok {
+			className, hasName := cluster.additionalInfo[clusterTopologyNameKey]
+			classNamespace, hasNamespace := cluster.additionalInfo[clusterTopologyNamespaceKey]
+			isOwner := false
+
+			if hasNamespace && hasName {
+				if className == clusterClass.identity.Name && clusterClass.identity.Namespace == classNamespace {
+					cluster.addSoftOwner(clusterClass)
+					isOwner = true
+				}
+			} else if hasName {
 				if className == clusterClass.identity.Name && clusterClass.identity.Namespace == cluster.identity.Namespace {
 					cluster.addSoftOwner(clusterClass)
+					isOwner = true
 				}
 			}
+
+			if !isOwner {
+				continue
+			}
+
+			// Set shouldNotDelete if clusterClass already referenced by some cluster
+			if _, found := ownerClasses[clusterClass]; found || clusterClass.identity.Namespace != cluster.identity.Namespace {
+				clusterClass.shouldNotDelete = true
+				clusterClass.shouldNotDeleteHierarchy = true
+			} else {
+				ownerClasses[clusterClass] = empty{}
```

Comment:
If I got it right, we do not delete a CC when it is referenced by more than one cluster.
If yes, I don't think this is correct, the number of cluster referencing a CC should not influence this decision

Reply:
Yeah, makes sense. Fixed now to only account for external to the moved group references.